src/vulkan-renderer/world/collision.cpp

```diff
@@ -0,0 +1,183 @@
+    m_nearest_corner = other.m_nearest_corner;
+    m_nearest_edge = other.m_nearest_edge;
+}
+
+template <typename T>
+RayCubeCollision<T>::RayCubeCollision(const T &cube, const glm::vec3 ray_pos, const glm::vec3 ray_dir) : m_cube(cube) {
+
+    // This lambda adjusts the center points on a cube's face to the size of the octree,
+    // so collision works with cubes of any size. This does not yet account for rotations!
+    const auto adjust_coordinates = [=](const glm::vec3 pos) {
+        // TODO: Take rotation of the cube into account.
+        return m_cube.center() + pos * (m_cube.size() / 2);
+    };
+
+    // x: left/right, y: front/back, z: top/bottom.
+    static constexpr std::array BBOX_DIRECTIONS{
+        glm::vec3(-1.0f, 0.0f, 0.0f), // left
+        glm::vec3(1.0f, 0.0f, 0.0f),  // right
+        glm::vec3(0.0f, -1.0f, 0.0f), // front
+        glm::vec3(0.0f, 1.0f, 0.0f),  // back
+        glm::vec3(0.0f, 0.0f, 1.0f),  // top
+        glm::vec3(0.0f, 0.0f, -1.0f)  // bottom
+    };
+
+    // The coordinates of the center of every face of the cube.
+    const std::array bbox_face_centers{adjust_coordinates(BBOX_DIRECTIONS[0]),  // left
+                                       adjust_coordinates(BBOX_DIRECTIONS[1]),  // right
+                                       adjust_coordinates(BBOX_DIRECTIONS[2]),  // front
+                                       adjust_coordinates(BBOX_DIRECTIONS[3]),  // back
+                                       adjust_coordinates(BBOX_DIRECTIONS[4]),  // top
+                                       adjust_coordinates(BBOX_DIRECTIONS[5])}; // bottom
+
+    const std::array bbox_corners{adjust_coordinates({-1.0f, -1.0f, -1.0f}), // left front bottom
+                                  adjust_coordinates({-1.0f, -1.0f, 1.0f}),  // left front top
+                                  adjust_coordinates({-1.0f, 1.0f, -1.0f}),  // left back bottom
+                                  adjust_coordinates({-1.0f, 1.0f, 1.0f}),   // left back top
+                                  adjust_coordinates({1.0f, -1.0f, -1.0f}),  // right front bottom
+                                  adjust_coordinates({1.0f, -1.0f, 1.0f}),   // right front top
+                                  adjust_coordinates({1.0f, 1.0f, 1.0f}),    // right back bottom
+                                  adjust_coordinates({1.0f, 1.0f, -1.0f})};  // right back top
+
+    using bbox_corner_on_face_index = std::array<std::size_t, 4>;
+
+    // These indices specify which 4 points describe the corners on the given face.
+    static constexpr std::array BBOX_CORNERS_ON_FACE_INDICES{
+        bbox_corner_on_face_index{0, 1, 2, 3}, // left
+        bbox_corner_on_face_index{4, 5, 6, 7}, // right
+        bbox_corner_on_face_index{0, 1, 4, 5}, // front
+        bbox_corner_on_face_index{2, 3, 6, 7}, // back
+        bbox_corner_on_face_index{1, 3, 5, 7}, // top
+        bbox_corner_on_face_index{0, 2, 4, 6}  // bottom
+    };
```

Comment:
Because this is a direct property of a Cube, I think it would be better to move this to the `Cube`-class.
Also, maybe use another name, e.g. `CORNER_INDICES_ON_FACE` or just `FACE_CORNERS`.

Reply:
Done.

Reply:
https://github.com/inexorgame/vulkan-renderer/issues/386